Issue 1044 - Implementation of the $everything operation

luisgarcc opened the pull request:
This PR implements the [FHIR $everything operation](https://www.hl7.org/fhir/operation-patient-everything.html).

The commits here create the new project for the new Patient `$everything` operation and register it with the necessary build & deployment artifacts.

An implementation of the `$everything` operation that obtains all subresources from the `CopmartmentUtil.getCompartmentResourceTypes()` and searches for all of the resources from those resource types related to a patient using `doSearch()`. 

The implementation includes support for the various query parameters in the operation definition (`start`, `end`, `_since`, `_type`, `_count`), that operate as permitted by the search service.

The commit includes the corresponding unit and functional tests.

Addresses issue #1044

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+/*
+ * (C) Copyright IBM Corp. 2021
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
Please use the SPDX version 

/*
 * (C) Copyright IBM Corp. 2021
 *
 * SPDX-License-Identifier: Apache-2.0
 */

luisgarcc replied:
Done

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+        }
+    }
+    
+    /* (non-Javadoc)
+     * @see com.ibm.fhir.server.operation.spi.AbstractOperation#buildOperationDefinition()
+     */
```
```suggestion
```

luisgarcc replied:
Done.

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+        }
+    }
+
+    /* (non-Javadoc)
+     * @see com.ibm.fhir.server.operation.spi.AbstractOperation#doInvoke(com.ibm.fhir.server.operation.spi.FHIROperationContext, java.lang.Class, java.lang.String, java.lang.String, com.ibm.fhir.model.resource.Parameters, com.ibm.fhir.server.operation.spi.FHIRResourceHelpers)
+     */
```
```suggestion
```

luisgarcc replied:
Done

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+        List<Entry> allEntries = new ArrayList<>(DEFAULT_RESOURCE_COUNT * 2);
+        allEntries.add(patientEntry);
+        
+        // We can't always use the "date" query parameter to query by clinical date, only with some resources.
+        // Initial list obtained from the github issue: https://github.com/IBM/FHIR/issues/1044#issuecomment-769788097
+        // Otherwise the search throws an exception. We create a params map with and without and use as needed
+        MultivaluedMap<String, String> queryParameters = parseQueryParameters(parameters);
+        MultivaluedMap<String, String> queryParametersWithoutDates = new MultivaluedHashMap<String, String>(queryParameters);
+        queryParametersWithoutDates.remove(DATE_QUERY_PARAMETER);
```
Parameters parameters and is mediate as a post or get. 
I suggest using the Parameters object as you don't have to jointly care about the input method.

luisgarcc replied:
@prb112 not getting this can you elaborate?

prb112 replied:
All of your query parameters are validated at this point and pushed into Parameters parameters object. 
You should probably access this object to get the queryParameters (or more directly the Parameters value) 

Look at FHIROperationUtil to see how it is generated.

luisgarcc replied:
I am using the validated `Parameter` from the `Parameters` object at this point to extract the parameters of the $everything operation and mapping those to the corresponding `MultivalueMap` query parameters that the `doSearch()` method will understand. Let me know if anything other than that is necessary. Thanks!

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+
+    private static final Logger LOG = java.util.logging.Logger.getLogger(EverythingOperation.class.getName());
+    
+    /**
+     * The <a href="https://www.hl7.org/fhir/search.html#prefix">prefix</a> used to indicate the start date for the $everything resources
+     */
+    protected static final String STARTING_FROM = "ge";
+
+    /**
+     * The <a href="https://www.hl7.org/fhir/search.html#prefix">prefix</a> used to indicate the end date for the $everything resources
+     */
+    protected static final String UP_UNTIL = "le";
```
Use the SearchConstants.Prefix.GE. and LE

luisgarcc replied:
Done

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+     */
+    protected static final String DATE_QUERY_PARAMETER = "date";
+
+    /**
+     * The "_lastUpdated" query parameter used in the underlying search operation.
+     */
+    protected static final String LAST_UPDATED_QUERY_PARAMETER = "_lastUpdated";
+
+    /**
+     * The query parameter to indicate a maximum resource count for the $everything operation
+     */
+    protected static final String COUNT_QUERY_PARAMETER = "_count";
+
+    /**
+     * The query parameter to indicate the resource types for the $everything operation
+     */
+    protected static final String TYPE_QUERY_PARAMETER = "_type";
+
```
should deduplicate with SearchConstants

luisgarcc replied:
I didn't see a specific constant for `_lastUpdated` but I added one.
As far as _count and _type, done.

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+     */
+    protected static final String SINCE_QUERY_PARAMETER = "_since";
+
+    /**
+     * The default max number of resources to search for each subsresource
+     */
+    protected static final int DEFAULT_RESOURCE_COUNT = 500;
```
This should be maximum 1000 and come from the SearchConstants class

luisgarcc replied:
Done.

prb112 commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,342 @@
+     * @return the list of patient subresources that will be included in the $everything operaetion
+     * @throws FHIRSearchException 
+     */
+    private List<String> getDefaultIncludedResourceTypes() throws FHIRSearchException {
+        List<String> resourceTypes = new ArrayList<>(CompartmentUtil.getCompartmentResourceTypes(PATIENT));
+        // TODO: Practitioner and Organization are not included in the getCompartmentReourceTypes() by default but it seems
+        // like a couple of good additional resources to include and they are even mentioned as examples of resources
+        // to include in the docs: https://www.hl7.org/fhir/operation-patient-everything.html 
+        // resourceTypes.add(Practitioner.class.getSimpleName());
+        // resourceTypes.add(Organization.class.getSimpleName());
+        return resourceTypes;
+    }
```
You should only really need this one time.

luisgarcc replied:
Yeah, I call this method from the constructor only.

prb112 commented on operation/fhir-operation-everything/pom.xml:
```diff
@@ -0,0 +1,32 @@
+            <artifactId>fhir-server</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <!-- Needed for MultiValuedMap -->
+        <dependency>
+            <groupId>jakarta.ws.rs</groupId>
+            <artifactId>jakarta.ws.rs-api</artifactId>
+        </dependency>
```
scope should be provided

luisgarcc replied:
Done.

prb112 commented on fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:
```diff
@@ -2845,8 +2845,8 @@ private String getRequestBaseUri(String type) throws Exception {
-        // Strip any path segments for whole-system interactions (in case of whole-system search, "Resource" is passed as the type)
-        if (type == null || type.isEmpty() || "Resource".equals(type)) {
+        // Strip any path segments for whole-system interactions (in case of whole-system search, "Resource" is passed as the type, or $everything-based search)
+        if (type == null || type.isEmpty() || "Resource".equals(type) || baseUri.contains("$everything")) {
```
is this being trimmed for a reason?

luisgarcc replied:
Yeah, otherwise the resulting `fullUrl` of the patient resources resulting from the search will have *$everything* in there too.

michaelwschroeder commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+     * @param parameters the operation parameters
+     * @return the {@link MultivaluedMap} for the search service built from the parameters 
+     */
+    protected MultivaluedMap<String, String> parseQueryParameters(Parameters parameters) {
```
May want to look at (or call) the validation done in AbstractOperation.validateInputParameters() - provides good detail on errors. Based on what mode you're in (strict vs. lenient - see my comment in getOverridenIncludeResourceTypes()), either just log the error or throw an exception.

luisgarcc replied:
I am not doing any validation here, if I understand correctly the validation already happened by the time we get here. I am just converting `$everything` query parameters to `$search` query parameters here, maybe I am missing something?

michaelwschroeder replied:
My bad - you're correct.

michaelwschroeder commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+        } else {
+            queryParameters.add(SearchConstants.COUNT, countParameter.getValue().as(FHIR_INTEGER).getValue() + "");
+        }
+        // We use gt/lt here in an effort to be more liberal in terms of what we return
```
The _count parameter is problematic in that you can't take advantage of the persistence layer's paging since you're making a separate call per resource type. If you're not going to implement paging at this layer, I would make that clear in the documentation, and just set _count to MAX_PAGE_SIZE on all your calls to FHIRRestHelper.doSearch().

luisgarcc replied:
Correct paging is not supported but they can provide a _count, I thought this matched the spec.

michaelwschroeder replied:
I think this works ok. If we're not supporting paging and we interpret _count as the max to be returned for the entire bundle, then we should never get more than _count resources per resource type.

michaelwschroeder commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+        for (String compartmentType : resourceTypes) {
+            MultivaluedMap<String, String> searchParameters = queryParameters;
+            if (!SUPPORT_CLINICAL_DATE_QUERY.contains(compartmentType)) {
+                searchParameters = queryParametersWithoutDates;
```
Would be nice to add a log msg here if the request specified the date parameter but it's not valid for this resource type.

luisgarcc replied:
Sure done.

michaelwschroeder commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+            }
+            Bundle results = null;
+            try {
+                results = resourceHelper.doSearch(compartmentType, PATIENT, logicalId, searchParameters, null, null, null);
```
If you're getting ALL results for a resource type, will have to call doSearch in a loop. Since there's a max page size the persistence layer will return, results might come back paged.
If there's a hard limit that you won't exceed, need to check here and throw an exception if exceeded.

luisgarcc replied:
I am not sure I completely follow, but I am not getting "all" resources of a type only the first 1000. 
This initial implementation was based on https://github.com/IBM/FHIR/issues/1044#issuecomment-769788097 saying that paging was not supported initially.

luisgarcc replied:
Done.

michaelwschroeder commented on operation/fhir-operation-everything/src/test/java/com/ibm/fhir/operation/everything/EverythingOperationTest.java:
```diff
@@ -0,0 +1,142 @@
+
+/**
+ * 
+ * @author Luis A. García
```
style - remove author tag

michaelwschroeder commented on operation/fhir-operation-everything/src/test/java/com/ibm/fhir/operation/everything/EverythingOperationTest.java:
```diff
@@ -0,0 +1,142 @@
+public class EverythingOperationTest {
+
+    
+    private EverythingOperation exportOperation;
```
Confusing having the operation named exportOperation
```suggestion
    private EverythingOperation everythingOperation;
```

luisgarcc replied:
done

michaelwschroeder commented on operation/fhir-operation-everything/src/test/java/com/ibm/fhir/operation/everything/EverythingOperationTest.java:
```diff
@@ -0,0 +1,142 @@
+     * 
+     */
+    @Test
+    public void testExportOperation() {
```
Same as above
```suggestion
    public void testEverythingOperation() {
```

luisgarcc replied:
done

michaelwschroeder commented on operation/fhir-operation-everything/src/test/java/com/ibm/fhir/operation/everything/EverythingOperationTest.java:
```diff
@@ -0,0 +1,142 @@
+            return FHIRParser.parser(Format.JSON).parse(reader);
+        }
+    }
+}
```
Maybe add some error condition tests

luisgarcc replied:
If I test passing in bad input here, I am not really testing my code which wouldn't even be reached, I would be testing the code listed here which I assume has already been unit tested
![image](https://user-images.githubusercontent.com/1807320/110180238-a464d400-7dc6-11eb-9bcd-b528f49a87b0.png)

JohnTimm commented:
You should be able to look up the operation definition in the FHIR registry instead of loading it from file
https://github.com/IBM/FHIR/blob/ed1c85e4f659824a1d4b6fd4ef264486353ece0d/operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java#L142-L150

luisgarcc commented on fhir-server-test/src/test/java/com/ibm/fhir/server/test/BundleTest.java:
```diff
@@ -2454,7 +2454,14 @@ private void printBundle(String method, String bundleType, Bundle bundle) throws
+     * @param expectedType
+     * @param expectedEntryCount
+     */
+    public static void assertResponseBundle(Bundle bundle, BundleType expectedType, int expectedEntryCount) {
```
Done.

luisgarcc commented on fhir-server-test/src/test/java/com/ibm/fhir/server/test/everything/EverythingOperationTest.java:
```diff
@@ -0,0 +1,300 @@
+
+/**
+ * 
+ * @author Luis A. García
```
Done all.

luisgarcc commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+            try {
+                results = resourceHelper.doSearch(compartmentType, PATIENT, logicalId, searchParameters, null, null, null);
+            } catch (Exception e) {
+                LOG.warning("Error retrieving $everything resources of type " + compartmentType + " for patient " + logicalId);
```
On this one I figured it would be better to return something than nothing, so that is why I err'd on this side. I can just let the exception go, but I'd like to hear other opinions. 

The product feels more robust/resilient this way, the problem is properly logged and in the event that this happens for say 1 type out of 20 where that 1 type may have only a few instances I think ppl would appreciate at least getting something. If they are really really missing some type they could check the logs and I could add something about using bulkexport instead? I could also print a trace for better problem determination?

luisgarcc commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+    /**
+     * @return the list of patient subresources that will be included in the $everything operation, as provided by the user
+     * @throws FHIRSearchException 
+     */
```
done

luisgarcc commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+        String typeOverridesParam = typesParameter.getValue().as(FHIR_STRING).getValue();
+        String[] typeOverridesList = typeOverridesParam.split(SearchConstants.JOIN_STR);
+        for (String typeOverride : typeOverridesList) {
+            typeOverrides.add(typeOverride.trim());
```
nice...done

luisgarcc commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+        Entry patientEntry = Entry.builder()
+                .resource(patient)
+                .fullUrl(patientURL)
+                .build();
```
done

luisgarcc commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,309 @@
+            throw exceptionWithIssue;
+        }
+        if (patient == null) {
+            FHIROperationException exceptionWithIssue = buildExceptionWithIssue("Patient with ID " + logicalId + " does not exist.", IssueType.NOT_FOUND);
```
done

luisgarcc commented on fhir-server-test/src/test/java/com/ibm/fhir/server/test/everything/EverythingOperationTest.java:
```diff
@@ -0,0 +1,300 @@
+            if (resources.remove(resourceId)) {
+                println("Found expected " + resourceType + ": " + resourceId);
+            } else {
+                println("Found unkown " + resourceType + ": " + resourceId);
```
done

michaelwschroeder commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,329 @@
+        }
+        FHIRRequestContext requestContext = FHIRRequestContext.get();
+        if (!HTTPHandlingPreference.LENIENT.equals(requestContext.getHandlingPreference()) && !unknownTypes.isEmpty()) {
+            throw new FHIRSearchException("The following types are not known: " + String.join(",", unknownTypes));
```
The types might be known, just not supported for patient compartment. Might be better to say "The following resource types are not supported by this operation: ". In the exception, should set an issue type of INVALID - see SearchExceptionUtil.buildInvalidException() - this will translate to a 4xx HTTP code rather than a 500 code. Also, if in lenient mode, we should log the message.

luisgarcc replied:
Done.

luisgarcc commented:
> 
> 
> You should be able to look up the operation definition in the FHIR registry instead of loading it from file
> 
> https://github.com/IBM/FHIR/blob/ed1c85e4f659824a1d4b6fd4ef264486353ece0d/operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java#L142-L150
> 
> You can see some examples of this in the `fhir-operation-term` project

@JohnTimm I realize this is possible, I just used the approach that @lmsurpre pointed me too in the `fhir-operation-document`. I can go either way, but it kinda feels like loading it locally may be more reliable in that there is no possibility of the FHIR the http://hl7.org server going down affecting loading up of a FHIR server's operation going up at that time. Again, I can go either way  but since there are other loading it locally I figured I would do that.

luisgarcc commented on fhir-server-test/src/test/java/com/ibm/fhir/server/test/BundleTest.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * (C) Copyright IBM Corp. 2017, 2020
+ * (C) Copyright IBM Corp. 2017, 2020, 2021
```
Done.

luisgarcc commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,329 @@
+        for (String compartmentType : resourceTypes) {
+            MultivaluedMap<String, String> searchParameters = queryParameters;
+            if (!SUPPORT_CLINICAL_DATE_QUERY.contains(compartmentType)) {
+                LOG.finest("The request specified the a '" + START_QUERY_PARAMETER + "' and/or '" + END_QUERY_PARAMETER + "' query parameter they are not valid for resource type '" + compartmentType + "'.");
```
done

michaelwschroeder commented on operation/fhir-operation-everything/src/main/java/com/ibm/fhir/operation/everything/EverythingOperation.java:
```diff
@@ -0,0 +1,335 @@
+        for (String compartmentType : resourceTypes) {
+            MultivaluedMap<String, String> searchParameters = queryParameters;
+            if (startOrEndProvided  && !SUPPORT_CLINICAL_DATE_QUERY.contains(compartmentType)) {
+                LOG.finest("The request specified the a '" + START_QUERY_PARAMETER + "' and/or '" + END_QUERY_PARAMETER + "' query parameter. They are not valid for resource type '" + compartmentType + "', so will be ignored.");
```
```suggestion
                LOG.finest("The request specified a '" + START_QUERY_PARAMETER + "' and/or '" + END_QUERY_PARAMETER + "' query parameter. They are not valid for resource type '" + compartmentType + "', so will be ignored.");
```

luisgarcc replied:
Done

luisgarcc commented:
@michaelwschroeder the latest commit here addresses the concerns you had raised so far as our discussion in the meeting 3/11/21.

Specifically the commit includes:

- Ignoring the `count` attribute and logging when used
- Failing the overall `$everything` operation if any of its searches fails
- Retrieving all resources of a compartment type by pagination, not just the first 1000

Thanks for your feedback,

lmsurpre commented:
This looks great to me.  The only thing I see missing is documentation.
@luisgarcc do you want to try your hand at adding some docs?

1. We try keeping a list of modules in our README, so would be good to list this under https://github.com/ibm/fhir#extended-operations.
2. We document which "extended operations" we support at https://ibm.github.io/FHIR/Conformance/#extended-operations.  You can click `Edit this page on GitHub` on the bottom of that page to get to the source (which is just in the `/docs` directory of this same repo).  This would be an especially good spot to document that we don't support `_since` at this time (or paging for the `$everything` operation in general).

michaelwschroeder reviewed (approved):
LGTM. Nice job Luis!

luisgarcc commented:
> 
> 
> This looks great to me. The only thing I see missing is documentation.
> @luisgarcc do you want to try your hand at adding some docs?

Sure!
> 
>     1. We try keeping a list of modules in our README, so would be good to list this under https://github.com/ibm/fhir#extended-operations.
> 

Done.

>     2. We document which "extended operations" we support at https://ibm.github.io/FHIR/Conformance/#extended-operations.  You can click `Edit this page on GitHub` on the bottom of that page to get to the source (which is just in the `/docs` directory of this same repo).  This would be an especially good spot to document that we don't support `_since` at this time (or paging for the `$everything` operation in general) and that there's a 10,000 resource limit.

Done.

Thank you.

lmsurpre commented on docs/src/pages/Conformance.md:
```diff
@@ -313,6 +313,7 @@ Type operations are invoked at `[base]/[resourceType]/$[operation]`
 | [$export](https://hl7.org/fhir/uv/bulkdata/OperationDefinition-patient-export.html) | Patient | Obtain a set of resources pertaining to all patients | exports to an S3-compatible data store; see the [user guide](https://ibm.github.io/FHIR/guides/FHIRServerUsersGuide/#410-bulk-data-operations) for config info |
 | [$document](https://hl7.org/fhir/R4/operation-composition-document.html) | Composition | Generate a document | Prototype-level implementation |
 | [$apply](https://hl7.org/fhir/R4/operation-plandefinition-apply.html) | PlanDefinition | Applies a PlanDefinition to a given context | A prototype implementation that performs naive conversion |
+| [$everything](https://www.hl7.org/fhir/operation-patient-everything.html) | Patient | Obtain all resources pertaining to a patient | Current implementation supports obtaining all resources for a patient up to an aggregate total of 10,000 resources (at which point it is recommended to use the `$export` operation). This implementation does not currently support using the `_since` and `count` quer ...
```
```suggestion
| [$everything](https://www.hl7.org/fhir/operation-patient-everything.html) | Patient | Obtain all resources pertaining to a patient | Current implementation supports obtaining all resources for a patient up to an aggregate total of 10,000 resources (at which point it is recommended to use the `$export` operation). This implementation does not currently support using the `_since` and `_count` query parameters. Pagination is not currently supported. |
```

luisgarcc replied:
Done

lmsurpre reviewed (approved):
LGTM
